### PR TITLE
Pipeline: use notify instead of polling for `ExchangeSender`

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #8869

Problem Summary:

### What is changed and how it works?

use `waitForWritable()` to replace `isWritable()`
```
    auto res = writer->waitForWritable();
    switch (res)
    {
    case WaitResult::Ready:
        return OperatorStatus::NEED_INPUT;
    case WaitResult::WaitForPolling:
        return OperatorStatus::WAITING;
    case WaitResult::WaitForNotify:
        return OperatorStatus::WAIT_FOR_NOTIFY;
    }
```

### Check List

Tests 

Tested tpch50, no performance regression

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by SeaRise** [comment]
I don't understand why there's always an error saying "Error: found files NOT formatted dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp," even though I've already formatted it locally.

**Comment by SeaRise** [comment]
> I don't understand why there's always an error saying "Error: found files NOT formatted dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp," in `pull-integration-test` even though I've already formatted it locally.

Thanks a lot, @Lloyd-Pottiger .

File: dbms/src/Flash/Executor/PipelineExecutorContext.cpp [hunk 1]
```diff
@@ -155,6 +176,12 @@ void PipelineExecutorContext::cancel()
     if (is_cancelled.compare_exchange_strong(origin_value, true, std::memory_order_release))
     {
         cancelSharedQueues();
+        if (likely(dag_context))
+        {
+            // Cancel the tunnel_set here to prevent pipeline tasks waiting in the WAIT_FOR_NOTIFY state from never being notified.
+            if (dag_context->tunnel_set)
```
Comment (windtalker) [inline dbms/src/Flash/Executor/PipelineExecutorContext.cpp:182]: The tunnels will be always closed at the destruct of mpp task, is that too late so you need to close it here?

File: dbms/src/Flash/Mpp/MPPTask.cpp [hunk 2]
```diff
@@ -759,13 +759,13 @@ void MPPTask::abort(const String & message, AbortType abort_type)
...
-            /// abort the components from top to bottom because if bottom components are aborted
-            /// first, the top components may see an error caused by the abort, which is not
+            /// abort mpptunnels first because if others components are aborted
+            /// first, the mpptunnels may see an error caused by the abort, which is not
             /// the original error
             setErrString(message);
             abortTunnels(message, false);
-            abortQueryExecutor();
             abortReceivers();
+            abortQueryExecutor();
```
Comment (windtalker) [inline dbms/src/Flash/Mpp/MPPTask.cpp:768]: why abort receiver first

File: dbms/src/Flash/Executor/PipelineExecutorContext.cpp (hunk 1 above)
Comment (SeaRise) [reply dbms/src/Flash/Executor/PipelineExecutorContext.cpp:182]: Yes, if a pipeline task is in the wait_for_notify state, and another pipeline task throws an error and cannot terminate the query, then abortMPPTunnel cannot be called, this is a deadlock.

Comment (windtalker) [reply dbms/src/Flash/Executor/PipelineExecutorContext.cpp:182]: OK

File: dbms/src/Operators/ExchangeSenderSinkOp.cpp [hunk 3]
```diff
@@ -39,14 +39,28 @@ OperatorStatus ExchangeSenderSinkOp::writeImpl(Block && block)
...
 
+OperatorStatus ExchangeSenderSinkOp::waitForWriter() const
+{
+    auto res = writer->waitForWritable();
+    switch (res)
+    {
+    case WaitResult::Ready:
+        return OperatorStatus::NEED_INPUT;
+    case WaitResult::WaitForPolling:
+        return OperatorStatus::WAITING;
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Operators/ExchangeSenderSinkOp.cpp:50]: yes, you are right.
Here only wait for tunnel ready will return `WaitForPolling`.

**Review by windtalker — APPROVED** [review]
LGTM

File: dbms/src/Common/GRPCQueue.h [hunk 4]
```diff
@@ -139,6 +139,9 @@ class GRPCSendQueue
 
     bool isWritable() const { return send_queue.isWritable(); }
 
+    void registerPipeReadTask(TaskPtr && task) { send_queue.registerPipeReadTask(std::move(task)); }
```
Comment (gengliqi) [inline dbms/src/Common/GRPCQueue.h:142]: Seems these `registerPipeReadTask`s are not used in this PR?

Comment (SeaRise) [reply dbms/src/Common/GRPCQueue.h:142]: Yes, they are used in the exchange receiver PR, but not in this PR.
https://github.com/pingcap/tiflash/pull/9073/files

**Review by gengliqi — APPROVED** [review]

**Pull request merged**